ntp-client/src/Network/NTP/Client.hs

```diff
@@ -71,248 +64,221 @@ data NtpStatus =
-              (threadDelay 30000000)
-              (do
-                atomically $ do
-                  s <- readTVar ncStatus
-                  case s of
-                    NtpSyncPending -> return ()
-                    _              -> retry
-                traceWith tracer NtpTraceSpawnNtpClientResolvePending
-              )
-            resolve
-          else return addrs
-
--- Try to create IPv4 and IPv6 socket.
-mkSockets :: Tracer IO NtpTrace -> NtpClientSettings -> IO Sockets
-mkSockets tracer settings =
-    doMkSockets `catch` handleIOException >>= \case
-        Just socks -> pure socks
-        Nothing     -> do
-            traceWith tracer NtpTraceMkSocketsNoSockets
---            logWarning "Couldn't create either IPv4 or IPv6 socket, retrying in 5 sec..."
-            threadDelay 5000000
-            mkSockets tracer settings
+        runProtocol _version _localAddr [] = return []
+        runProtocol _version Nothing    _  = return []
+        runProtocol version (Just addr) servers = do
+            socketAction tracer ntpSettings addr servers >>= \case
+                Left err -> do
+                    traceWith tracer $ NtpTraceRunProtocolError version err
+                    return []
+                Right [] -> do
+                    traceWith tracer $ NtpTraceRunProtocolNoResult version
+                    return []
+                Right r@(_:_) -> do
+                    traceWith tracer $ NtpTraceRunProtocolSuccess version
+                    return r
+
+socketAction ::
+       Tracer IO NtpTrace
+    -> NtpClientSettings
+    -> AddrInfo
+    -> [AddrInfo]
+    -> IO (Either IOError [ReceivedPacket])
+socketAction tracer netSettings localAddr destAddrs
+    = tryIOError $ bracket acquire release action
   where
-    doMkSockets :: IO (Maybe Sockets)
-    doMkSockets = do
-        addrs <- udpLocalAddresses
-        (<>) <$> (createAndBindSock tracer IPv4 addrs)
-             <*> (createAndBindSock tracer IPv6 addrs)
-
-    handleIOException :: IOException -> IO (Maybe Sockets)
-    handleIOException e = do
-        traceWith tracer $ NtpTraceMkSocketsIOExecption e
-        threadDelay 5000000
-        doMkSockets
+    acquire :: IO Socket
+    acquire = do
+        s <- Socket.socket (addrFamily localAddr) Datagram Socket.defaultProtocol
+        traceWith tracer NtpTraceSocketOpen
+        return s
+
+    release :: Socket -> IO ()
+    release s = do
+        Socket.close s
+        traceWith tracer NtpTraceSocketClosed
+
+    action :: Socket -> IO [ReceivedPacket]
+    action socket = do
+        Socket.setSocketOption socket ReuseAddr 1
+        Socket.bind socket (addrAddress localAddr)
+        inQueue <- atomically $ newTVar []
+        _err <- withAsync (send socket  >> loopForever)      $ \sender ->
+               withAsync timeout    $ \delay ->
+               withAsync (reader socket inQueue ) $ \revc ->
+                    waitAnyCancel [sender, delay, revc]        
```

Comment:
I thought we agreed with @semanticphilosopher that we can pipeline the requests and then await for responses.  This would required only two threads (including one for timeout one).

The indentation and spacing is wired.

Reply:
I like this better.
I think it reflects better what we actually want and it is not more complicated either.
sendManyTo is not expected to block, but it does not harm to have it covered by a timeout.
Also there is the implicit assumption that we time-stamp the incoming packets as fast as possible (or more precise there is the assumption that the transmission times for a request/response are the same in both directions)
If we wait for all requests to be send, it will skew the measurement.
(And also there is now the 100ms rate-limiting when sending.)

Reply:
That's ok.